src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
                 _isReferenced = true;
             }
 
-            private void AddItemsToCache(ImmutableHashSet<string> globsToIgnore, ImmutableList<ItemData> items)
+            private void AddItemsToCache(ImmutableHashSet<string> globsToIgnore, OrderedItemDataCollection items)
             {
                 if (_cache == null)
                 {
-                    _cache = new Dictionary<ISet<string>, ImmutableList<ItemData>>();
+                    _cache = new Dictionary<ISet<string>, OrderedItemDataCollection>();
                 }
 
                 _cache[globsToIgnore] = items;
             }
         }
 
+        /// <summary>
+        /// A collection of ItemData that maintains insertion order and internally optimizes some access patterns, e.g. bulk removal
+        /// based on normalized item values.
+        /// </summary>
+        internal sealed class OrderedItemDataCollection
+        {
+            #region Inner types
+
+            /// <summary>
+            /// An efficient multi-value wrapper holding one or more versioned items.
+            /// </summary>
+            internal struct DictionaryValue : IEnumerable<I>
+            {
+                /// <summary>
+                /// The version of the containing collection at the time this value was last changed.
+                /// </summary>
+                private int _version;
+
+                /// <summary>
+                /// Holds one value or a list of values.
+                /// </summary>
+                private object _value;
+
+                public DictionaryValue(int version, I item)
+                {
+                    _version = version;
+                    _value = item;
+                }
+
+                public void Add(int version, I item)
+                {
+                    if (_value is List<I> list)
+                    {
+                        if (version != _version)
+                        {
+                            list = new List<I>(list);
```

Comment:
If Add gets called a lot on the same instance on the hot path the list constructor may end up creating a lot of arrays right? Switching to ImmutableList would alleviate that.

Profiling OrchardCore would probably show if this is an issue or not.

Reply:
The list is needed for cases where the same item (or, more precisely, an item with the same path-normalized value) is included multiple times.

```xml
  <ItemGroup>
    <A Include="file" />    
    <A Include="file" />    
  </ItemGroup>
```

When later we evaluate a remove that matches `file`, both items should be removed from the running list. To support immutability, we keep track of the collection version and clone the list when it would mutate an older version.

This shouldn't be common, only ~80 lists are created during a clean build of OC.

I went with a plain List for perf because this cloning is super rare (duplicate items + memoization) but you're right that `ImmutableList` is exactly what I need and simplifies the code a bit.

Reply:
Switched to `ImmutableList`, thank you!